File: src/core/Image.ts
import ImageType from './ImageType.js'
import type TypedArray from './TypedArray.js'
import setMatrixElement from './setMatrixElement.js'
import Metadata from './Metadata.js'

class Image {
  name: string = 'image'

  origin: number[]

  spacing: number[]

  direction: TypedArray

  size: number[]

  metadata: Metadata

  data: null | TypedArray

  constructor (public readonly imageType = new ImageType()) {
    const dimension = imageType.dimension
    this.origin = new Array(dimension)
    this.origin.fill(0.0)

    this.spacing = new Array(dimension)
    this.spacing.fill(1.0)

    this.direction = new Float64Array(dimension * dimension)
    this.direction.fill(0.0)
    for (let ii = 0; ii < dimension; ii++) {
      setMatrixElement(this.direction, dimension, ii, ii, 1.0)
    }

    this.size = new Array(dimension)
    this.size.fill(0)

    this.metadata = new Map()

    this.data = null
  }
}

export default Image

File: packages/core/java/src/main/java/org/itk/wasm/Image.java
package org.itk.wasm;

import java.util.*;

public class Image {
	public ImageType imageType;
	public String name;
	public List<Double> origin;
	public List<Double> spacing;
	public double[][] direction;
	public List<Integer> size;
	public Map<String, Object> metadata;
	public double[] data;

	public Image() {
		this.imageType = new ImageType();
		this.name = "Image";
		this.origin = new ArrayList<>();
		this.spacing = new ArrayList<>();
		this.direction = new double[0][0];
		this.size = new ArrayList<>();
		this.metadata = new HashMap<>();
		this.data = null;
	}

	public void postInit() {
		if (imageType instanceof Map) {
			imageType = new ImageType();
			Map<String, Object> imageTypeMap = asdict(imageType);
			// Set values from the map to the corresponding fields in ImageType
			// Example: imageType.setDimension((int) imageTypeMap.get("dimension"));
			// Add similar code for other fields
		}

		int dimension = imageType.dimension;
		if (origin.isEmpty()) {
			for (int i = 0; i < dimension; i++) {
				origin.add(0.0);
			}
		}

		if (spacing.isEmpty()) {
			for (int i = 0; i < dimension; i++) {
				spacing.add(1.0);
			}
		}

		if (direction.length == 0) {
			direction = new double[dimension][dimension];
			for (int i = 0; i < dimension; i++) {
				for (int j = 0; j < dimension; j++) {
					if (i == j) {
						direction[i][j] = 1.0;
					} else {
						direction[i][j] = 0.0;
					}
				}
			}
		}

		if (size.isEmpty()) {
			for (int i = 0; i < dimension; i++) {
				size.add(1);
			}
		}
	}

	// Add getters and setters for the fields
}
